src/FetchWorkerTool.js

```diff
@@ -0,0 +1,191 @@
+ */
+class PrivateFetchWorkerTool {
+    constructor () {
+        /**
+         * What does the worker support of the APIs we need?
+         * @type {{fetch:boolean}}
+         */
+        this._workerSupport = {
+            fetch: typeof fetch !== 'undefined'
+        };
+
+        /**
+         * A possible error occurred standing up the worker.
+         * @type {!Error}
+         */
+        this._supportError = null;
+
+        /**
+         * The worker that runs fetch and returns data for us.
+         * @type {!Worker}
+         */
+        this.worker = null;
+
+        /**
+         * A map of ids to fetch job objects.
+         * @type {object}
+         */
+        this.jobs = {};
+
+        try {
+            if (this.getSupported) {
+                // eslint-disable-next-line global-require
+                const FetchWorker = require('worker-loader?{"inline":true,"fallback":true}!./FetchWorkerTool.worker');
+
+                this.worker = new FetchWorker();
+
+                this.worker.addEventListener('message', ({data}) => {
+                    if (data.support) {
+                        this._workerSupport = data.support;
+                        return;
+                    }
+                    for (const message of data) {
+                        if (this.jobs[message.id]) {
+                            if (message.error) {
+                                this.jobs[message.id].reject(message.error);
+                            } else {
+                                this.jobs[message.id].resolve(message.buffer);
+                            }
+                            delete this.jobs[message.id];
+                        }
+                    }
+                });
+            }
+        } catch (error) {
+            this._supportError = error;
+        }
+    }
+
+    /**
+     * Is get supported?
+     *
+     * false if the environment does not workers, fetch, or fetch from inside a
+     * worker. Finding out the worker supports fetch is asynchronous and will
+     * guess that it does if the window does until the worker can inform us.
+     * @returns {boolean} Is get supported?
+     */
+    get getSupported () {
+        return (
+            typeof Worker !== 'undefined' &&
+            this._workerSupport.fetch &&
+            !this._supportError
+        );
+    }
+
+    /**
+     * Request data from a server with a worker using fetch.
+     * @param {{url:string}} reqConfig - Request configuration for data to get.
+     * @param {{method:string}} options - Additional options to configure fetch.
+     * @returns {Promise.<Buffer>} Resolve to Buffer of data from server.
+     */
+    get ({url}, options = {method: 'GET'}) {
+        // TODO: This is rather knowledgeable of what data urls look like. Can
+        // it be abstracted or less knowledgable while letting FetchTool do the
+        // job with data?
+        if (/\/\w+\/get\/$|\.json/.test(url)) {
```

Comment:
What is the behavior of this if e.g., someone is using the scratch gui with different project/asset servers configured with a different URL structure? I think it's a high priority to avoid having this kind of deep knowledge if it's possible.

Reply:
I'm going to update WebHelper to choose a tool based on assetType (one for Projects and one for everything else).